py/server/deephaven/table.py

```diff
@@ -1282,14 +1282,25 @@ def count_by(self, col: str, by: Union[str, Sequence[str]] = None) -> Table:
         except Exception as e:
             raise DHError(e, "table count_by operation failed.") from e
 
-    def agg_by(self, aggs: Union[Aggregation, Sequence[Aggregation]], by: Union[str, Sequence[str]] = None) -> Table:
+    def agg_by(self, aggs: Union[Aggregation, Sequence[Aggregation]], by: Union[str, Sequence[str]] = None,
+               preserve_empty: bool = False, initial_groups: Table = None) -> Table:
         """The agg_by method creates a new table containing grouping columns and grouped data. The resulting
         grouped data is defined by the aggregations specified.
 
         Args:
             aggs (Union[Aggregation, Sequence[Aggregation]]): the aggregation(s)
-            by (Union[str, Sequence[str]]): the group-by column name(s), default is None
-
+            by (Union[str, Sequence[str]]): the group-by column name(s), if not provided, all rows from this table are
+                grouped into a single group of rows before the aggregations are applied to the result, default is None.
+            preserve_empty (bool): whether to keep result rows for groups that are initially empty or become empty as
+                a result of updates. Each aggregation operator defines its own value for empty groups. Default is False.
+            initial_groups (Table): a table whose distinct combinations of values for the group-by column name(s)
+                should be used to create an initial set of aggregation groups. All other columns are ignored. This is
+                useful in combination with preserve_empty == True to ensure that particular groups appear in the result
+                table, or with preserve_empty == False to control the encounter order for a collection of groups and
+                thus their relative order in the result. Changes to this table are not expected or handled; if this
+                table is a refreshing table, only its contents at instantiation time will be used. Default is None,
+                the result will be the same as if a table is provided but no rows were supplied. When it is provided,
+                the 'by' argument must be provided to explicitly specify the grouping columns.
```

Comment:
`preserve_empty` is not tri-state. `None` had better be treated as `False` or an error. The Java Table API only supports a primitive `boolean` for this argument.

Isn't the stuff you're reacting to in the section for `initial_groups`? Is it really that confusing?

- Encounter order is the output order for all aggregations; output rows are ordered relative to the order in which the operation first observed them, so if a group "appears" earlier in upstream row key space on a subsequent update it still shows up after existing groups in the downstream output.
- It's *slightly* different than the norm, because it's the only case in the whole query engine where you might get something other than strict, repeatable order preservation; we do this for practical reasons, because of the impact on downstream queries.